ignite/metrics/loss.py

```diff
@@ -22,8 +22,8 @@ def reset(self):
 
     def update(self, output):
         y_pred, y = output
-        loss_sum = self._loss_fn(Variable(y_pred, volatile=True), Variable(y, volatile=True), size_average=False)
-        self._sum += loss_sum.data[0]
+        average_loss = self._loss_fn(Variable(y_pred, volatile=True), Variable(y, volatile=True))
+        self._sum += average_loss.data[0] * y.shape[0]
```

Comment:
Maybe we can avoid this with:
```
self._sum += average_loss.data[0]
self._num_examples += 1
```

Reply:
@vfdev-5 Unfortunately we need to do it this way as we can't guarantee that all batches will be the same size.

Reply:
> we can't guarantee that all batches will be the same size.
This is sounds strange to me. @jasonkriss could you hint me an example, please?

My reasoning is that `self._loss_fn` does the average itself as `nn.NLLLoss()`. Maybe it is always the case.

Reply:
if your dataset has 1000 examples and you are doing batches of size 256, then your last batch will be of size 232. Therefore to get the true mean loss, we have to make sure we take into account the number of examples in the batch.

Reply:
@jasonkriss  I agree with what you say. Anyway, `y.shape[0]` assumes that `y` is `Tensor` and this may be a problem with multi-output...

Reply:
@vfdev-5 once #121 is merged, that won't be an issue any longer.

Reply:
@jasonkriss imagine a case when `self._loss_fn` is a sort of composite loss and works on `y_pred` and `y` being a list of tensors, so no need of `output_transform` (from #121) to apply, and `y.shape[0]` will fail...

Reply:
I think at that point, the recommended path should probably just be a custom `Metric` or subclass of `Loss`. Unless you can think of a simple way to support it here.

Reply:
Going back to 
```python
self._sum += average_loss.data[0] * y.shape[0]
self._num_examples += y.shape[0]
```
what if my criterion is `nn.NLLLoss(size_average=False)` (for whatever reason), then this will fail. 

I mean to support all cases we can get super verbose in the code i.e.
```python
if isinstance(loss_fn, nn.Loss):
    if loss_fn.size_average:
      .....
     else:
       .....
else:
      self.loss_fn(y_pred, y, size_average=False)
      .....
```